src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -2052,14 +2052,14 @@ private BoundExpression BindNonMethod(SimpleNameSyntax node, Symbol symbol, Bind
 
             bool isUsedBeforeDeclaration(SimpleNameSyntax node, LocalSymbol localSymbol)
             {
-                Location localSymbolLocation = localSymbol.GetFirstLocation();
+                var declarator = localSymbol.GetDeclaratorSyntax();
 
-                if (node.SyntaxTree == localSymbolLocation.SourceTree)
-                {
-                    return node.SpanStart < localSymbolLocation.SourceSpan.Start;
-                }
+                // trivial position check, before more costly tree check (which requires walking up the nodes). Most
+                // code is correct, so this check is expected to succeed nearly every time.
+                if (node.SpanStart >= declarator.SpanStart)
+                    return false;
 
-                return false;
+                return node.SyntaxTree == declarator.SyntaxTree;
```

Comment:
Here, I am wondering how we would possibly think this `node` might be a use of `localSymbol`, unless it is already in the same syntax tree. No need to attempt to change logic as part of eliminating Location usages, though.